File: src/scripts/slidingWindow.lua
-- Sliding Window Lua Script
local key = KEYS[1]
local now = tonumber(ARGV[1])      -- current time in ms
local window = tonumber(ARGV[2])   -- window size in ms
local limit = tonumber(ARGV[3])    -- max requests allowed in window
local memberId = ARGV[4]           -- unique request identifier

local clearBefore = now - window

-- Remove elements outside the window
redis.call("ZREMRANGEBYSCORE", key, 0, clearBefore)

-- Count current requests in window
local currentRequests = redis.call("ZCARD", key)

local allowed = 0
if currentRequests < limit then
    redis.call("ZADD", key, now, memberId)
    allowed = 1
    currentRequests = currentRequests + 1
end

-- Expose TTL to clear unused sliding windows
local ttlSeconds = math.ceil(window / 1000) * 2
if ttlSeconds < 3600 then
    ttlSeconds = 3600
end
redis.call("EXPIRE", key, ttlSeconds)

local remaining = limit - currentRequests
return { allowed, remaining }

File: src/strategies/slidingWindow.js
// src/strategies/slidingWindow.js
import { BaseRateLimiterStrategy } from "./base.js";
import { redis, isRedisConnected } from "../config/redis.js";
import { LRUCache } from "lru-cache";
import { logger } from "../utils/logger.js";

export class SlidingWindowStrategy extends BaseRateLimiterStrategy {
    constructor(options = {}) {
        super(options);
        // In-memory fallback cache storing lists of request timestamps
        this.fallbackCache = new LRUCache({
            max: options.maxFallbackItems || 10000,
            ttl: 1000 * 60 * 60, // 1 hour default TTL
        });
    }

    async allowRequest(key, rule) {
        const { limit, windowMs } = rule;
        const nowMs = Date.now();
        const uniqueId = `${nowMs}-${Math.random().toString(36).substring(2, 11)}`;

        if (isRedisConnected()) {
            try {
                // evalSlidingWindow returns [allowed, remaining]
                const result = await redis.evalSlidingWindow(
                    key,
                    nowMs,
                    windowMs,
                    limit,
                    uniqueId
                );

                const allowed = result[0] === 1;
                const remaining = parseInt(result[1], 10);
                const resetTime = Math.ceil((nowMs + windowMs) / 1000);

                return {
                    allowed,
                    remaining,
                    resetTime,
                };
            } catch (err) {
                logger.error(`Redis Sliding Window failed, falling back to memory: ${err.message}`);
            }
        }

        // In-memory fallback evaluation
        let timestamps = this.fallbackCache.get(key) || [];
        const clearBefore = nowMs - windowMs;

        // Filter out timestamps outside current window
        timestamps = timestamps.filter(t => t > clearBefore);

        let allowed = false;
        let remaining = limit - timestamps.length;

        if (timestamps.length < limit) {
            timestamps.push(nowMs);
            allowed = true;
            remaining = limit - timestamps.length;
        }

        this.fallbackCache.set(key, timestamps);

        // Reset time is when the oldest timestamp within the window will slide out
        const oldestTimestamp = timestamps.length > 0 ? timestamps[0] : nowMs;
        const resetTime = Math.ceil((oldestTimestamp + windowMs) / 1000);

        return {
            allowed,
            remaining,
            resetTime,
        };
    }
}
